### ISSUE TITLE
All element not supported as child of extension

### ISSUE THREAD

**Issue by schlagert** [willemdj/erlsom#34]
Hi,

I'm currently trying to compile the XSD definitions provided in [RFC 5707](https://www.ietf.org/rfc/rfc5707.txt). Unfortunately `erlsom` doesn't like `all` elements directly contained in `extension` elements. Here's an example taken from the `msml-conf-core-datatypes.xsd`:

```xsd
  <xs:element name="modifyconference" substitutionGroup="msmlRequest">
    <xs:complexType>
      <xs:complexContent>
        <xs:extension base="msmlRequestType">
          <xs:all>
            <xs:element name="audiomix" type="audioMixType" minOccurs="0"/>
            <xs:element name="videolayout" type="videoLayoutType" minOccurs="0"/>
          </xs:all>
          <xs:attribute name="id" type="confID.datatype" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
  </xs:element>
```

A minimal example from the RFC would consist of the two files `msml-core-datatypes.xsd` and `msml-conf-core-datatypes.xsd` (which I would have attached to the issue if that would be possible).

Adding the `#alt{tag = 'xsd:all', tp = 'allType'}` alternative to the `extensionTypeC` in `erlsom_parseXsd.erl` seems to solve the issue. However, I don't know if that is enough to add support for this structure.

Regards
Tobias

**Comment by willemdj** [comment]
Hello Tobias,

I'll have a look. Do I understand correctly from your email that you tried
the solution that you are proposing, and that it works for you?

Regards,
Willem

**Comment by schlagert** [comment]
Hi Willem,

when I apply this small patch, I'm able to compile the XSD's provided by the RFC and create header files from them. However, I don't know if the patch is sufficient to be able to use the model for data binding or XML document creation.

If this should be enough I could support you by issuing a corresponding PR.

Regards
Tobias

**Comment by willemdj** [comment]
Hi Tobias,

I did some basic tests and it appears to work. If you issue a pull request I will pull it.

Regards,
Willem

**Comment by schlagert** [comment]
Using PR #35 enables Erlsom to compile the XSD provided by the RFC.

Unfortunately, this reveals a few more problems:
 * Duplicate schema inclusions are not handled by Erlsom (addressed by PR #36).
 * The generated `.hrl` does not compile (when included). There seems to be a problem with `type` and `name` name collisions.
 * The type hierarchy of the XSD is weird and leads to duplicate type references in `choice` fields, e.g.

  ```erlang
choice :: 'msmlRequestTypes/send'() | msmlRequestType() | destroyconference() | modifyconference() | 'msmlRequestTypes/send'() | modifyconference() | destroyconference(),
  ```

**Comment by willemdj** [comment]
Hello Tobias,

There are options that will allow you to specify a prefix for types and
groups, see the documentation for erlsom:compile_xsd. Using (one of) these
options should solve your issue with the compilation of the .hrl file (and
possibly also the duplicate reference in the choice field, I am not sure).

Regards,
Willem

Option  = {prefix, Prefix} |
          {type_prefix, TypePrefix} |
          {group_prefix, GroupPrefix} |
          {include_fun, Include_fun} |
          {include_dirs, Include_dirs} |
          {include_files, Include_files}

**Comment by schlagert** [comment]
Hi Willem,

unfortunately, I couldn't make it work (I've already tried the different `prefix` options).

As I've read in other forums, the XSD also exposes some really rare corner-cases like the *unstable name conflict* making it hard to interpret. For more information on that refer to this [thread](http://www.codesynthesis.com/pipermail/xsd-users/2007-November/001332.html).

In the meantime I gave good old `xmerl` a try and after fixing another bug there (see erlang/otp#808) it is at least able to validate the XSD against XML documents. It would have been nice to have data binding but I can also live with *manual parsing*, so I'll stick with it for now.

Neverthess, Erlsom is a cool project and I will definitely use it for other projects, don't worry about this complex and crappy corner-case XSD! Regarding pull requests #35 and #36: Although, these are only small changes, I hope you will still find them useful.

Regards,
Tobias
